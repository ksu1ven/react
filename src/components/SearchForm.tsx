import React from 'react';
import { useRouter } from 'next/router';
import { useRef, useState } from 'react';
import { updateQueryParams } from '../utils/helpFunctions';

function SearchForm(props: Record<'searchValue', string>) {
  const { searchValue } = props;
  const inputCurrentValue = useRef(searchValue);
  const [errorOccured, setErrorOccured] = useState(false);

  const router = useRouter();

  if (errorOccured) {
    throw new Error("Hello, I'm Error!");
  }
  return (
    <form
      action="#"
      className="w-3/5 flex justify-center m-auto gap-x-10 rounded"
      data-testid="search-form"
      onSubmit={(e) => {
        e.preventDefault();
        localStorage.setItem('search', inputCurrentValue.current);
        if (searchValue !== inputCurrentValue.current)
          router.push(
            '?' +
              updateQueryParams(
                router.query,
                'search',
                inputCurrentValue.current
              )
          );
      }}
    >
      <button
        type="button"
        className="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
        onClick={() => {
          setErrorOccured(true);
        }}
      >
        Error
      </button>

      <input
        defaultValue={searchValue}
        type="text"
        name="search"
        id="search"
        data-testid="search-input"
        className="w-1/3 p-3"
        onChange={(e) => {
          e.target.value = e.target.value.trim();
          inputCurrentValue.current = e.target.value;
        }}
      />
      <button
        type="submit"
        className="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
      >
        Search
      </button>
    </form>
  );
}
export default SearchForm;
