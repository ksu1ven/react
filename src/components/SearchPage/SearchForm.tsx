import { useState, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import type { RootState } from '../../redux/store/store';
import { SetURLSearchParams } from 'react-router-dom';
import { setSearchValue } from '../../redux/features/searchSlice';
import {
  setPageNumber,
  setPaginationButtonsValue,
} from '../../redux/features/paginationSlice';
import { updateQueryParams } from '../../utils/helpFunctions';

interface Props {
  params: URLSearchParams;
  setParams: SetURLSearchParams;
}

function SearchForm(props: Props) {
  const searchValue = useSelector(
    (state: RootState) => state.search.searchValue
  );
  const pageNumber = useSelector(
    (state: RootState) => state.pagination.pageNumber
  );
  const dispatch = useDispatch();
  const inputCurrentValue = useRef(searchValue);

  const { params, setParams } = props;

  const [errorOccured, setErrorOccured] = useState(false);

  if (errorOccured) {
    throw new Error("Hello, I'm Error!");
  }
  return (
    <form
      action="#"
      className="w-3/5 flex justify-center m-auto gap-x-10 rounded"
      data-testid="search-form"
      onSubmit={(e) => {
        e.preventDefault();
        localStorage.setItem('searchValue', inputCurrentValue.current);
        dispatch(setSearchValue(inputCurrentValue.current));
        dispatch(setPaginationButtonsValue([1, 2, 3]));
        setParams(updateQueryParams(params, 'search', ''));
        if (pageNumber) dispatch(setPageNumber(0));
      }}
    >
      <button
        type="button"
        className="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
        onClick={() => {
          setErrorOccured(true);
        }}
      >
        Error
      </button>

      <input
        defaultValue={searchValue}
        type="text"
        name="search"
        id="search"
        data-testid="search-input"
        className="w-1/3 p-3"
        onChange={(e) => {
          e.target.value = e.target.value.trim();
          inputCurrentValue.current = e.target.value;
        }}
      />
      <button
        type="submit"
        className="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
      >
        Search
      </button>
    </form>
  );
}

export default SearchForm;
