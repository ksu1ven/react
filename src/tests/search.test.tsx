import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import SearchForm from '../components/SearchForm';

vi.mock('next/router', () => vi.importActual('next-router-mock'));

describe('Tests for the Search component', () => {
  it('Clicking the Search button saves the entered value to the local storage', async () => {
    mockRouter.push('/');
    render(<SearchForm searchValue="" />, {
      wrapper: MemoryRouterProvider,
    });
    expect(localStorage.getItem('search')).toBe(null);
    fireEvent.change(screen.getByTestId('search-input'), {
      target: { value: 'hotdog' },
    });
    fireEvent.submit(screen.getByTestId('search-form'));

    expect(mockRouter.asPath).toBe('/?search=hotdog');
    expect(localStorage.getItem('search')).toBe('hotdog');

    fireEvent.change(screen.getByTestId('search-input'), {
      target: { value: 'cat' },
    });
    fireEvent.submit(screen.getByTestId('search-form'));
    expect(screen.getByTestId('search-form')).toMatchSnapshot();

    expect(mockRouter.asPath).toBe('/?search=cat');
    expect(localStorage.getItem('search')).toBe('cat');
  });

  it('The component retrieves the value from the local storage upon mounting.', async () => {
    const localStorageValue = localStorage.getItem('search');
    expect(localStorageValue).toBe('cat');
    mockRouter.push(`/?search=${localStorageValue}`);
    render(
      <SearchForm searchValue={mockRouter.query.search?.toString() || ''} />,
      {
        wrapper: MemoryRouterProvider,
      }
    );

    expect(screen.getByTestId('search-form')).toMatchSnapshot();
    expect(screen.getByTestId('search-input')).toHaveValue(localStorageValue);
  });
});
