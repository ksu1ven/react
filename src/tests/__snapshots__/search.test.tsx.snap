// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Search component > Clicking the Search button saves the entered value to the local storage 1`] = `
<form
  action="#"
  class="w-3/5 flex justify-center m-auto gap-x-10 rounded"
  data-testid="search-form"
>
  <button
    class="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
    type="button"
  >
    Error
  </button>
  <input
    class="w-1/3 p-3"
    data-testid="search-input"
    id="search"
    name="search"
    type="text"
    value=""
  />
  <button
    class="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Tests for the Search component > The component retrieves the value from the local storage upon mounting. 1`] = `
<form
  action="#"
  class="w-3/5 flex justify-center m-auto gap-x-10 rounded"
  data-testid="search-form"
>
  <button
    class="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
    type="button"
  >
    Error
  </button>
  <input
    class="w-1/3 p-3"
    data-testid="search-input"
    id="search"
    name="search"
    type="text"
    value="cat"
  />
  <button
    class="w-1/5 bg-lime-700 p-3 rounded text-white font-extrabold"
    type="submit"
  >
    Search
  </button>
</form>
`;
